source/opt/constructs.h

```diff
@@ -0,0 +1,229 @@
+
+  SpvOp opcode() const { return opcode_; }
+  uint32_t type_id() const { return type_id_; }
+  uint32_t result_id() const { return result_id_; }
+  // Returns the first word of |index|th operand.
+  uint32_t GetSingleWordOperand(uint32_t index) const {
+    return GetSingleWordPayload(index + TypeResultIdCount());
+  }
+  uint32_t NumOperandWords() const;
+  // Returns the payload at the |index|th operand.
+  const Payload& GetOperand(uint32_t index) const {
+    return GetPayload(index + TypeResultIdCount());
+  }
+  uint32_t NumOperands() const {
+    return static_cast<uint32_t>(payloads_.size() - TypeResultIdCount());
+  }
+  // Returns the first word of |index|th payload.
+  uint32_t GetSingleWordPayload(uint32_t index) const;
+  uint32_t NumPayloadWords() const {
+    return NumOperandWords() + TypeResultIdCount();
+  }
+  // Returns the |index|th payload.
+  const Payload& GetPayload(uint32_t index) const;
+  uint32_t NumPayloads() const {
+    return static_cast<uint32_t>(payloads_.size());
+  }
+  void SetPayload(uint32_t index, std::vector<uint32_t>&& data);
+  std::vector<Inst>& dbg_line_info() { return dbg_line_info_; }
+
+  bool IsNop() const {
+    return opcode_ == SpvOpNop && type_id_ == 0 && result_id_ == 0 &&
+           payloads_.empty();
+  }
+  void ToNop() {
+    opcode_ = SpvOpNop;
+    type_id_ = result_id_ = 0;
+    payloads_.clear();
+  }
+
+  // Pushes the binary segments for this instruction into the back of *|binary|.
+  // If |keep_nop| is true and this is a OpNop, avoids pushing.
+  void ToBinary(std::vector<uint32_t>* binary, bool keep_nop) const;
+
+ private:
+  uint32_t TypeResultIdCount() const {
+    return (type_id_ != 0) + (result_id_ != 0);
+  }
+
+  SpvOp opcode_;
+  uint32_t type_id_;
+  uint32_t result_id_;
+  std::vector<Payload> payloads_;
+  // Opline and OpNoLine instructions preceding this instruction. Note that for
+  // Insts representing OpLine or OpNonLine itself, this field should be empty.
+  std::vector<Inst> dbg_line_info_;
+};
+
+class BasicBlock {
+ public:
+  BasicBlock(Function* function, Inst&& label)
+      : function_(function), label_(label) {}
+
+  void AddInstruction(Inst&& i) { insts_.push_back(std::move(i)); }
+
+  void ForEachInst(const std::function<void(Inst*)>& f);
+
+  // Pushes the binary segments for this instruction into the back of *|binary|.
+  // If |keep_nop| is true and this is a OpNop, avoids pushing.
+  void ToBinary(std::vector<uint32_t>* binary, bool keep_nop) const;
+
+ private:
+  Function* function_;
+  Inst label_;
+  std::vector<Inst> insts_;
+};
+
+class Function {
+ public:
+  Function(Module* module, Inst&& def_inst)
+      : module_(module), def_inst_(def_inst), end_inst_(SpvOpFunctionEnd) {}
+
+  void AddParameter(Inst&& p) { params_.push_back(std::move(p)); }
+  void AddBasicBlock(BasicBlock&& b) { blocks_.push_back(std::move(b)); }
+
+  void ForEachInst(const std::function<void(Inst*)>& f);
```

Comment:
Why not just give const ref access to the underlying vector?  Are you wanting to act on the label as well?

needs blurbs.